### PR TITLE
Fixed esexporter serialized document to avoid conflicting keys

### LINKED ISSUE #33264: [exporter/elasticsearch] Regression: Duplicate Key in JSON After #31694
### Component(s)

exporter/elasticsearch

## Description
Previously, the request to ELK contained a well-structured JSON object with unique keys:

```
{
  ...
  "host": {
    "name": "hostname"
  },
  "os": {
    "description": "Windows 11 10.0",
    "type": "windows"
  },
  ...
}
```
However, after the change, the JSON format has been modified, resulting in a duplicate host key:
```
{
  ...
  "host": {
    "hostname": "hostname",
    "os": {
      "full": "Windows 11 10.0",
      "platform": "windows"
    }
  },
  ...
  "host": {  // Duplicate key
    "os": {
      "type": "windows"
    }
  },
  ...
}
```

This duplicate key renders the JSON invalid, leading to parsing errors from ELK:

```
{
  "error": {
    "reason": "failed to parse",
    "type": "mapper_parsing_exception"
  },
  "status": 400
}
```
## Steps to Reproduce
add the following resource attribute to OTLP request
host.name: Str(hostname)
os.description: Str(Windows 11 10.0)
os.type: Str(windows)

## Expected Result
ELK successfully parses the JSON data and processes it without errors.

## Actual Result
ELK encounters a parsing error due to the duplicate host key, and the data is not processed successfully.

### Collector version

0.101.0

## Environment
Not relevant

### OpenTelemetry Collector configuration

```yaml
exporters:
  elasticsearch/logs:
    ...
    mapping:
      mode: ecs
    ....
```

### Log output

```shell
2024-05-27T06:13:19.135Z	debug	elasticsearchexporter@v0.101.0/elasticsearch_bulk.go:45	Request roundtrip completed.	{"kind": "exporter", "data_type": "logs", "name": "elasticsearch/logs", "path": "/_bulk", "method": "POST", "duration": 0.057272404, "status": "200 OK"}
2024-05-27T06:13:19.136294606Z 2024-05-27T06:13:19.136Z	error	elasticsearchexporter@v0.101.0/elasticsearch_bulk.go:318	Drop docs: failed to index: struct { Type string "json:\"type\""; Reason string "json:\"reason\"" }{Type:"mapper_parsing_exception", Reason:"failed to parse"}	{"kind": "exporter", "data_type": "logs", "name": "elasticsearch/logs", "status": 400}
2024-05-27T06:13:19.136315906Z github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter.(*worker).flush
2024-05-27T06:13:19.136322306Z 	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter@v0.101.0/elasticsearch_bulk.go:318
2024-05-27T06:13:19.136327106Z github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter.(*worker).run
2024-05-27T06:13:19.136332306Z 	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter@v0.101.0/elasticsearch_bulk.go:304
2024-05-27T06:13:19.136337506Z github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter.newBulkIndexer.func1
2024-05-27T06:13:19.136342306Z 	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter@v0.101.0/elasticsearch_bulk.go:220
```

### Additional context

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31694#issuecomment-2132746471

### PR DESCRIPTION
**Pull request by michalpristas**
**Description:** 
`Sort()` or `Dedup()` call was missing before runnin `Serialize()` with ECS mapping enabled

**Link to tracking Issue:** 
Fixes #33264 

**Testing:** Added UT

**Documentation:** -

### CODE REVIEW

File: exporter/elasticsearchexporter/model_test.go [hunk 1]
```diff
@@ -218,6 +218,61 @@ func TestEncodeEvents(t *testing.T) {
...
+		"service.name":     "foo.bar",
+		"service.version":  "1.1.0",
+		"agent.name":       "otlp",
+		"host.os.platform": "darwin",
+		"host.os.full":     "Mac OS Mojave",
+		"host.os.name":     "Mac OS X",
+		"host.os.version":  "10.14.1",
+		"host.os.type":     "macos",
+		"host.name":        "localhost",
+	})
```
Comment (ycombinator) [inline exporter/elasticsearchexporter/model_test.go:245]: Maybe I'm missing something obvious, but is `expectedDocFields` being used in this test?

File: exporter/elasticsearchexporter/model_test.go [hunk 2]
```diff
@@ -218,6 +218,61 @@ func TestEncodeEvents(t *testing.T) {
...
+		"event.name": "user-password-change",
+	})
+	require.NoError(t, err)
+	observedTimestamp := pcommon.Timestamp(1710273641123456789)
+	record.SetObservedTimestamp(observedTimestamp)
+
+	m := encodeModel{
+		mode:  MappingECS,
+		dedot: true,
+		dedup: true,
```
Comment (ycombinator) [inline exporter/elasticsearchexporter/model_test.go:267]: How is this test testing de-duplication?  I ran it with setting `dedup: false` and it still passes.

File: exporter/elasticsearchexporter/model_test.go (hunk 1 above)
Comment (michalpristas) [reply exporter/elasticsearchexporter/model_test.go:245]: yes they are not. leftovers

File: exporter/elasticsearchexporter/model_test.go (hunk 2 above)
Comment (michalpristas) [reply exporter/elasticsearchexporter/model_test.go:267]: if you run it without a fix, resulting document is malformed. 
the fix was really in calling `Sort` or `Dedup` before `Serailize`

**Review by carsonip — COMMENTED** [review]
Thanks, can you also undo the change in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32917 ?

**Comment by carsonip** [comment]
> Link to tracking Issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33264

Do you mean "Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33264" ?

**Review by ycombinator — APPROVED** [review]

**Comment by michalpristas** [comment]
do we need changelog for this one?

**Review by carsonip — APPROVED** [review]

**Review by andrzej-stencel — APPROVED** [review]
If I understand correctly, the fix is to sort the document fields before serializing when `dedot` is enabled (the default).

**Pull request merged**